ironfish/src/wallet/wallet.ts

```diff
@@ -747,15 +741,65 @@ export class Wallet {
     }
   }
 
-  async createSpends(
+  private buildAmountsNeeded(
+    receives: {
+      publicAddress: string
+      amount: bigint
+      memo: string
+      assetIdentifier: Buffer
+    }[],
+    burns: BurnDescription[],
+    fee: bigint,
+  ): BufferMap<bigint> {
+    const amountsNeeded = new BufferMap<bigint>()
+
+    amountsNeeded.set(Asset.nativeIdentifier(), fee)
+    for (const { amount, assetIdentifier } of receives) {
+      const currentAmount = amountsNeeded.get(assetIdentifier) ?? BigInt(0)
+      amountsNeeded.set(assetIdentifier, amount + currentAmount)
+    }
+
+    for (const { assetIdentifier, value } of burns) {
+      const currentAmount = amountsNeeded.get(assetIdentifier) ?? BigInt(0)
+      amountsNeeded.set(assetIdentifier, value + currentAmount)
+    }
+
+    return amountsNeeded
+  }
+
+  private async createSpends(
     sender: Account,
+    amountsNeeded: BufferMap<bigint>,
+  ): Promise<Array<{ note: Note; witness: NoteWitness }>> {
+    const notesToSpend: Array<{ note: Note; witness: NoteWitness }> = []
+
+    for (const [assetIdentifier, amountNeeded] of amountsNeeded.entries()) {
+      const { amount, notes } = await this.createSpendsForAsset(
+        sender,
+        assetIdentifier,
+        amountNeeded,
+      )
+      if (amount < amountNeeded) {
+        throw new NotEnoughFundsError(
+          `Insufficient funds: Needed ${amountNeeded.toString()} but have ${amount.toString()}`,
+        )
```

Comment:
Should we include the assetIdentifier in the message? Or even as an attribute of the error type?